### PR TITLE
Expose more options of `MultiBlockPlot3DReader` (`vtkMultiBlockPlot3DReader`)

### PR DESCRIPTION
**Pull request by whophil**
Expose options in `MultiBlockPlot3DReader` related to _functions_ - derived scalar/vector fields computed from the point data living in Plot3D files.

This PR adds an enum to map useful names to their int values required by the `add_function()` method. There may be a better place for this enum to live.

There are still many methods from the `vtkMultiBlockPlot3DReader` available I chose not to expose, mainly relating to the format of the target file. The VTK documentation states that most binary files can be read without calling these methods, so long as "auto-detect" is on - in PyVista, it is by default. Thus, for simplicity, I chose not to expose the many reading flags. A user can still set them directly on the VTK reader instance if necessary.

### CODE REVIEW

**Comment by MatthewFlamm** [comment]
PyVista doesn't have many (or any?) `Enum`s.  Now there are two PRs with #3115 .  Most selections for functions, like in this PR, have typically used strings, which isn't necessarily desirable.  But, I'm wary that using an `IntEnum` in this PR is going to break consistency.  We could consider using `StrEnum` for the existing string usage and for the one here, but it doesn't come until Python 3.11!  We can also inherit directly from `str` to get similar features as `StrEnum` in the meantime.

I think we should have some consistent usage across the code base, whether it is plain strings, `Enum`s that are all `StrEnum`-like, or `Enum`s that are various types.  One knock for `Enum` usage is that these objects will start clogging the namespace in PyVista.

If we decide to keep the `Enum` usage, can it be part of this class?  This way it wouldn't pollute the namespace and this one can only be used with this class.

**Comment by adeak** [comment]
> PyVista doesn't have many (or any?) `Enum`s.

For the record, I added `LightType` at https://github.com/pyvista/pyvista/blob/main/pyvista/plotting/lights.py#L18 via https://github.com/pyvista/pyvista/pull/1040.

**Comment by tkoyama010** [comment]
> We could consider using StrEnum for the existing string usage and for the one here, but it doesn't come until Python 3.11!

I think we can wait until 3.11 arrives and move it to `StrEnum` when it comes. I am +100 that namespace should be managed by str. However, I don't see a problem with IntEnum since VTK guarantees that it will not duplicate it. I think the current policy is to inherit VTK's Int as it is.

**Comment by whophil** [comment]
> If we decide to keep the `Enum` usage, can it be part of this class? This way it wouldn't pollute the namespace and this one can only be used with this class.

@MatthewFlamm as a nested class? Or something else?

**Comment by MatthewFlamm** [comment]
> > If we decide to keep the `Enum` usage, can it be part of this class? This way it wouldn't pollute the namespace and this one can only be used with this class.
> 
> @MatthewFlamm as a nested class? Or something else?

I was thinking the class could be stored as a class or instance attribute.  So, something like:

```
reader.add_function(MultiBlockPlot3DReader.FunctionEnum.DENSITY)
# or
reader.add_function(reader.FunctionEnum.DENSITY)
```

But having written it out, I'm not sure anymore.  It was more of a thought for discussion than a request for change.

**Comment by whophil** [comment]
@MatthewFlamm definitely worth the discussion.

> `MultiBlockPlot3DReader.FunctionEnum.DENSITY`

I believe this syntax implies a nested class. I don't know if we do that anywhere else in Pyvista. In general, I don't see it done much in any Python projects, but it does minimize the namespace clutter in this case.

```python
class MultiBlockPlot3DReader(BaseReader):
    """MultiBlock Plot3D Reader."""

        class FunctionEnum(enum.IntEnum):
            PRESSURE = 110
            ...
```

**Comment by adeak** [comment]
Just to plug my `Lights` again, we have the enum
https://github.com/pyvista/pyvista/blob/690acb8311b22943090091c4fe2283a5cab2f9bd/pyvista/plotting/lights.py#L18-L23

But then its values are pulled into the `Lights` class namespace:
https://github.com/pyvista/pyvista/blob/690acb8311b22943090091c4fe2283a5cab2f9bd/pyvista/plotting/lights.py#L122-L125

It would obviously be messier with a dozen enum values, but we could still do this to have `MultiBlockPlot3DReader.DENSITY` etc. Odds are the semantic scope of `MultiBlockPlot3DReader` is small enough that the enum values don't stand out too much.

**Comment by MatthewFlamm** [comment]
Having talked it through I think I agree that nesting it into the reader might be too eager.  I think the current form is fine since it doesn't clog up the top-level namespace.

File: pyvista/utilities/reader.py [hunk 1]
```diff
@@ -1248,6 +1249,30 @@ class Plot3DMetaReader(BaseReader):
     _class_reader = staticmethod(_vtk.lazy_vtkPlot3DMetaReader)
 
 
+class Plot3DFunctionEnum(enum.IntEnum):
```
Comment (MatthewFlamm) [inline pyvista/utilities/reader.py:1252]: Does this need to be added to the documentation somehow?  I think it will be missing, but I haven't built the docs to check for this PR.

Comment (MatthewFlamm) [inline pyvista/utilities/reader.py:1252]: Should we rename this to match more closely with the reader name?  Maybe it is too long though?

File: pyvista/utilities/reader.py [hunk 2]
```diff
@@ -1295,6 +1320,68 @@ def auto_detect_format(self):
     def auto_detect_format(self, value):
         self.reader.SetAutoDetectFormat(value)
 
+    def add_function(self, value: Union[int, Plot3DFunctionEnum]):
+        """Specify additional functions to read.
+
+        Parameters
+        ----------
+        value
```
Comment (MatthewFlamm) [inline pyvista/utilities/reader.py:1328]: I think we need a description of the mapping from int to Enum, or a direct link to the Enum documentation which might suffice.

File: pyvista/utilities/reader.py [hunk 3]
```diff
@@ -1295,6 +1320,68 @@ def auto_detect_format(self):
...
+            An integer or `Plot3DFunctionEnum`.
+        """
```
Comment (MatthewFlamm) [inline pyvista/utilities/reader.py:1330]: An example would be useful, in particular it won't be obvious that the user has to use `pyvista.reader.Plot3DFunctionEnum`.

File: pyvista/utilities/reader.py [hunk 4]
```diff
@@ -1295,6 +1320,68 @@ def auto_detect_format(self):
...
+            value = value.value
+        self.reader.AddFunction(value)
+
+    def remove_function(self, value: Union[int, Plot3DFunctionEnum]):
+        """Remove one function from list of functions to read.
+
+        Parameters
+        ----------
+        value
+            An integer or `Plot3DFunctionEnum`.
```
Comment (MatthewFlamm) [inline pyvista/utilities/reader.py:1341]: Same as above.

File: tests/utilities/test_reader.py [hunk 5]
```diff
@@ -384,22 +384,47 @@ def test_multiblockplot3dreader():
...
     reader.add_q_files(q_filename)
+    reader.add_function(112)  # add by int
+    reader.add_function(pyvista.reader.Plot3DFunctionEnum.PRESSURE_GRADIENT)  # add by enum
     mesh = reader.read()
     for m in mesh:
         assert len(m.array_names) > 0
 
+    assert 'MachNumber' in mesh[0].point_data
+    assert 'PressureGradient' in mesh[0].point_data
+
```
Comment (MatthewFlamm) [inline tests/utilities/test_reader.py:395]: Can we add tests for the remove functionality by int and Enum?

File: pyvista/utilities/reader.py (hunk 1 above)
Comment (adeak) [reply pyvista/utilities/reader.py:1252]: Just in case you're unaware, you can download the build artifact on successful CI doc builds via the checks tab -> build documentation (on the left) -> docs-build (under Artifacts), in this case https://github.com/pyvista/pyvista/suites/7764071910/artifacts/326430707

Comment (whophil) [reply pyvista/utilities/reader.py:1252]: The class name would be `MultiBlockPlot3DReaderFunctionEnum`. Happy to change it if that's preferred.

File: tests/utilities/test_reader.py (hunk 5 above)
Comment (whophil) [reply tests/utilities/test_reader.py:402]: Good catch, now added!

File: pyvista/utilities/reader.py (hunk 1 above)
Comment (MatthewFlamm) [reply pyvista/utilities/reader.py:1252]: Too long I think, let's keep the one you have.

**Comment by whophil** [comment]
Unfortunately, I can't build the docs locally on Mac to understand what is going on with the doc build.

I think it should be clear what I was _trying_ to achieve with the docs - creating a new section for reader enumerations, and then documenting the enumerations using [enum_tools](https://enum-tools.readthedocs.io/en/latest/).

I would appreciate help with the docs, CC @akaszynski @tkoyama010

**Comment by akaszynski** [comment]
> Unfortunately, I can't build the docs locally on Mac to understand what is going on with the doc build.
> 
> I think it should be clear what I was _trying_ to achieve with the docs - creating a new section for reader enumerations, and then documenting the enumerations using [enum_tools](https://enum-tools.readthedocs.io/en/latest/).

Fixed in 912bfc3044ba627abeab1b76e078c9509187448c.

**Comment by whophil** [comment]
Thanks @akaszynski ! After adding some numpydoc excludes, the docs are building now. However, the enum docs are added to the top level readers page, as shown below. Not sure if this is the preferred way of documenting this...

image

**Comment by tkoyama010** [comment]
Great documentation. I should follow it in #3150 .

**Comment by whophil** [comment]
I go back and forth on whether aliasing the enum members as class variables actually makes things clearer. I am open to suggestions.

**Comment by whophil** [comment]
Looking for some final feedback on:
- If we want to keep the enum members inside the reader class
- If the documentation for the enum is satisfactory

Thanks @MatthewFlamm @adeak @akaszynski !

**Comment by MatthewFlamm** [comment]
> Looking for some final feedback on:
> 
> * If we want to keep the enum members inside the reader class
> * If the documentation for the enum is satisfactory
> 
> Thanks @MatthewFlamm @adeak @akaszynski !

I'm not a user of this format, but it seems like these function settings would be an integral part of using this file format.  If that is an accurate representation then I am +1 on having them be aliased inside the reader class.  If it is a niche usage, I would swing slightly the other way.

File: requirements_docs.txt [hunk 6]
```diff
@@ -1,6 +1,7 @@
 Sphinx==4.5.0
 cmocean==2.0
 colorcet==3.0.0
+enum_tools
```
Comment (adeak) [inline requirements_docs.txt:4]: Just a drive-by comment:
1. Do we really need to add `enum_tools` and `sphinx-toolbox`? (Asking just because we should have very good reasons for adding additional dependencies, even if just for doc building.)
2. We should pin these dependencies too and let dependabot bump the versions as they are released.

Comment (akaszynski) [reply requirements_docs.txt:4]: I don't see where `sphinx-toolbox` is used. I'm fine with adding `enum_tools` provided we pin the dependency.

Comment (whophil) [reply requirements_docs.txt:4]: `sphinx-toolbox` appears to be an unlisted dependency of `enum_tools` (the docs build failed without it). I can pin the version numbers, though.

**Comment by whophil** [comment]
> I'm not a user of this format, but it seems like these function settings would be an integral part of using this file format. If that is an accurate representation then I am +1 on having them be aliased inside the reader class. If it is a niche usage, I would swing slightly the other way.

@MatthewFlamm in my opinion, the functions are likely to be used by anybody who reads in Plot3D files.

File: requirements_docs.txt (hunk 6 above)
Comment (tkoyama010) [reply requirements_docs.txt:4]: Now it seems v0.9.0 (What is 0.9.0.post1?).
```suggestion
enum-tools==0.9.0
```

**Review by akaszynski — APPROVED** [review]
Made one minor commit to improve coverage, but otherwise this LGTM.

**Review by MatthewFlamm — APPROVED** [review]
Also LGTM.

**Comment by akaszynski** [comment]
Thanks for rerunning the CI. I'll merge once it's green. Sorry, been busy with other PRs.

**Comment by whophil** [comment]
Thanks @akaszynski @MatthewFlamm @adeak !
